pkg/controllers/node/node_controller.go

```diff
@@ -52,31 +53,75 @@ type NodeController struct {
 }
 
 // NewNodeController Constructor for NodeController
-func NewNodeController(ctx context.Context, k8sClientset *kubernetes.Clientset, calicoClient client.Interface) controller.Controller {
+func NewNodeController(ctx context.Context, k8sClientset *kubernetes.Clientset, calicoClient client.Interface, cfg *config.Config) controller.Controller {
+	nc := &NodeController{
+		ctx:          ctx,
+		calicoClient: calicoClient,
+		k8sClientset: k8sClientset,
+		rl:           workqueue.DefaultControllerRateLimiter(),
+	}
 	// channel used to kick the controller into scheduling a sync. It has length
 	// 1 so that we coalesce multiple kicks while a sync is happening down to
 	// just one additional sync.
-	schedule := make(chan interface{}, 1)
+	nc.schedule = make(chan interface{}, 1)
 
 	// Create a Node watcher.
 	listWatcher := cache.NewListWatchFromClient(k8sClientset.CoreV1().RESTClient(), "nodes", "", fields.Everything())
 
-	// Informer handles managing the watch and signals us when nodes are deleted.
-	_, informer := cache.NewIndexerInformer(listWatcher, &v1.Node{}, 0, cache.ResourceEventHandlerFuncs{
+	// Setup event handlers
+	handlers := cache.ResourceEventHandlerFuncs{cache.ResourceEventHandlerFuncs{
 		DeleteFunc: func(obj interface{}) {
 			// Just kick controller to wake up and perform a sync. No need to bother what node it was
 			// as we sync everything.
-			kick(schedule)
-		},
-	}, cache.Indexers{})
+			kick(nc.schedule)
+		}}}
 
-	return &NodeController{
-		ctx:          ctx,
-		informer:     informer,
-		calicoClient: calicoClient,
-		k8sClientset: k8sClientset,
-		rl:           workqueue.DefaultControllerRateLimiter(),
-		schedule:     schedule,
+	// Only setup node labels syncer if env var is set
+	if cfg.SyncNodeLabels {
+		handlers.AddFunc = func(obj interface{}) {
+			nc.applyKddNodeLabels(obj.(*v1.Node))
+		}
+		handlers.UpdateFunc = func(_, obj interface{}) {
+			nc.applyKddNodeLabels(obj.(*v1.Node))
+		}
+	}
+
+	// Informer handles managing the watch and signals us when nodes are deleted.
+	//   also syncs up labels between k8s/calico node objects
+	_, nc.informer = cache.NewIndexerInformer(listWatcher, &v1.Node{}, 0, handlers, cache.Indexers{})
+
+	return nc
+}
+
+// applyKddNodeLabels applies the labels found in v1.Node and to calico's Node.
+func (nc *NodeController) applyKddNodeLabels(node *v1.Node) {
+	// Get calico's Node object.
+	calNode, err := nc.calicoClient.Nodes().Get(context.Background(), node.Name, options.GetOptions{})
+	if err != nil {
+		log.WithError(err).Error("failed to get node: " + node.Name)
```

Comment:
We need to do something in this case - we can log an error, but that's not really enough. We at _least_ need to return from this function, but really we should retry a few times before giving up.